OEL-1371: Project status and progress js

donquixote opened the pull request.

donquixote committed:
OEL-1371: Split and clean up preprocess functions for project.

donquixote committed:
OEL-1371: Add type hints on preprocess functions.

donquixote committed:
OEL-1371: Move part of project into content banner.

donquixote committed:
OEL-1371: Simplify code for badges in project content banner.

donquixote committed:
OEL-1371: Clear out empty badges in project content banner.

donquixote committed:
OEL-1371: Put project details id on the heading, drop the div below.

donquixote committed:
OEL-1371: Render only subgroups/children of group_project_details, not the group wrapper.

donquixote committed:
OEL-1371: Split the header macro.

donquixote committed:
OEL-1371: Use margin-top instead of padding-top for headings.

We need to use mt-4-5 instead of pt-3, to compensate for margin collapse.

donquixote committed:
OEL-1371: Absorb top margin if first element is not project details.

donquixote committed:
OEL-1371: Also use macro for project details header.

The top margin is ok, it will be absorbed.

donquixote committed:
OEL-1371: Add some blank lines for readability.

donquixote committed:
OEL-1371: Reorder imports.

donquixote committed:
OEL-1371: Add BCL component for project contributions.

donquixote committed:
OEL-1371: Use dynamic element for project status and progress.

donquixote committed:
OEL-1371: Rename twig variable 'inpage_navigation_fields' -> 'inpage_navigation_content'.

donquixote committed:
OEL-1371: Twig array keys in quotes.

donquixote committed:
OEL-1371: Temporarily set a shorter interval for testing. Revert this!

donquixote committed:
OEL-1371: Revert the shorter interval for testing.

donquixote committed:
OEL-1371: Fix padding-bottom on description lists with border-bottom.

donquixote committed:
OEL-1371: Hide the progress and status for users with js disabled.

This is a basic solution that will cause page jumping for users with js enabled.

donquixote committed:
OEL-1371: Remove pointless '|raw' for badges in content banner.

donquixote committed:
OEL-1371: Apply spaceless in project template.

donquixote committed:
OEL-1371: Drop pointless local variables in ContentProjectRenderTest.

donquixote committed:
OEL-1371: Use anonymous user instead of admin in ContentProjectRenderTest.

donquixote committed:
OEL-1371: Rename vars in ContentProjectRenderTest.

donquixote committed:
OEL-1371: Reuse the same instance of DescriptionListAssert in ContentProjectRenderTest.

donquixote committed:
OEL-1371: Reorder imports in ContentProjectRenderTest.

donquixote committed:
OEL-1371: Remove unused variable $page in ContentProjectRenderTest.

donquixote committed:
OEL-1371: Set up selenium.

donquixote committed:
OEL-1371: Test front-end javascript using WebDriverTestBase.

donquixote committed:
OEL-1371: Remove parts from ContentProjectRenderTest that are already done in the parent class.

donquixote committed:
OEL-1371: Insert placeholder values for non-js to make tests fail if js does not run.

donquixote committed:
OEL-1371: Initial javascript already runs before first assertions, no need for wait().

donquixote committed:
OEL-1371: Use getHtml(), because for some reason getText() returns empty string.

drishu commented on templates/content/node--oe-project--full.html.twig:
```diff
@@ -54,12 +88,14 @@
   {% endblock %}
 {% endset %}
+
 <div class="mt-md-4-75 mt-4">
```
to much whitespace now between the content and the hero, remove the div, maybe replace by `<article{{attributes}}>` up to you

drishu commented on modules/oe_whitelabel_extra_project/config/install/core.entity_view_display.node.oe_project.full.yml:
```diff
@@ -165,16 +163,6 @@ content:
-    third_party_settings: {  }
-    weight: 1
-    region: content
   oe_project_funding_programme:
```
Ups, I forgot to check this at last review, no skos should be linked
```
settings:
      link: false
```
this is valid for any skos field!

drishu commented on templates/content/node--oe-project--full.html.twig:
```diff
@@ -4,46 +4,80 @@
+{% set inpage_navigation_content %}
+  {#
+    Reliably absorb top margin of first element.
+    @todo Add a utility class 'eat-margin-top' with :before style.
```
was the todo discussed with FRONT colleagues? I'd skip the comment all together, up to you, but the way I see it, its something that we add to reflect bcl composition, so I dont feel the need to explain it further

donquixote replied:
In SRB I created a solution for this that uses utility classes and does not require an empty div.
I created an issue here, https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-1500

donquixote replied:
I think this div will look strange without a comment, and somebody will remove it eventually, not knowing what change this will cause.

drishu replied:
I don't agreed but not a blocker for me, we can keep it

drishu commented on templates/content/node--oe-project--full.html.twig:
```diff
@@ -4,46 +4,80 @@
+        'chart': true,
+        'legend': _contribution_legend_args,
+      } only %}
+    {% else %}
```
even if I don't fill out one field, I dont get into this else, lets drop it, and report to FRONT, they should take care of this situation on their side in the template

donquixote replied:
> even if I don't fill out one field, I dont get into this else 

strange. I do.

> they should take care of this situation on their side in the template

true.

drishu commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,135 @@
+    }
+  };
+
+})(bootstrap, Drupal, jQuery);
```
Something is wrong with the status, below it should be planned/not-started, see
<img width="1029" alt="Screenshot 2022-04-27 at 09 48 30" src="https://user-images.githubusercontent.com/11507647/165458360-7275278e-c02b-46d0-9235-77993add55af.png">
and here ongoing no?
<img width="1029" alt="Screenshot 2022-04-27 at 09 50 43" src="https://user-images.githubusercontent.com/11507647/165458597-584bb129-ae26-4848-b89d-e1d8b0f1ae9a.png">
Also the date formats displayed are wrong, should be what we use everywhere, like 22 February 2022

donquixote replied:
There seems to be a limit for window.setTimeout() due to 32 bit integer. If the timeout duration is higher than that number, it will happen immediately instead. The highest duration seems to be around 10 years.

Try in browser console: `window.setTimeout(console.log, 10 * 360 * 24 * 60 * 60 * 1000, '10 years passed'); window.setTimeout(console.log, 11 * 360 * 24 * 60 * 60 * 1000, '11 years passed');`.

See https://stackoverflow.com/questions/3468607/why-does-settimeout-break-for-large-millisecond-delay-values

Either way I will remove this stuff, too much risk and it was not asked for.

drishu commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,135 @@
+    const msNow = Date.now();
+    if (msNow < msBegin) {
+      setStatus(0);
+      window.setTimeout(setStatus, msBegin - msNow, 1);
```
I appreciate the intention, and viewing the bigger picture, but, a few things:
1. We don't have requirements for this, so we should not have done it in the first place
2. Scripting like this is prone to error `Uncaught TypeError: Cannot read properties of null (reading 'classList')`
3. Animating elements on the page is unexpected, is an accessibility issue, and should be limited to situations where an action has been performed (ex a progress bar in admin when a bulk action is performed, or rotating an icon when hovering or clicking on an accordion item)
4. we should leave such work to the FRONT and UX, that is their department and such things need to be requested and validated by business first in the UX side
5. this only adds unnecessary work and complexity, harder to maintain, harder to test
6. not worth, as we are dealing with days, not hours or minutes, so the effect is not even visible.
There was one more concern about usability but I forgot while writing up the rest.

The js should run only once, on page load, and set the appropriate status and progress depending on the time intervals. Thats it. Don't forget about KISS.

donquixote replied:
> We don't have requirements for this, so we should not have done it in the first place

True. My mistake.

> we should leave such work to the FRONT and UX, that is their department and such things need to be requested and validated by business first in the UX side

I would agree even without the animation, because now our stuff depends on selectors coming from front-end.
It was not my decision.

drishu commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,135 @@
+  ];
+
+  /**
+   * @param {int} msBegin
```
short desc is missing

drishu commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  /**
```
short desc

drishu commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,92 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  if ($date_range_item === NULL) {
+    return;
+  }
+  $status_keys = ['planned', 'ongoing', 'closed'];
```
not used

donquixote replied:
True. I will squash this into the other commit where I insert the placeholder values.
So that we get it back on revert.

drishu commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,92 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+    return;
+  }
+  $status_keys = ['planned', 'ongoing', 'closed'];
+  $status_labels = [t('Planned'), t('Ongoing'), t('Closed')];
```
think you can move this more down, and then you can merge the 2 ifs

drishu commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,92 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  // Project starts at the beginning of the first day at 00:00.
+  $t_start = $get_timestamp($date_range_item->value);
+  // Project ends at the end of the last day at 24:00.
+  $t_end = $get_timestamp($date_range_item->end_value . ' +1 day');
```
Not sure if its an issue or not (in this specific use case), because it passes to the next day in the calendar, so if end date is 31st it passes to 1st of next month. the difference is only 1 second but what I would have done is 
```suggestion
  $t_end = $get_timestamp($date_range_item->end_value . '  23:59:59');
```
this way you don't pass into the next day, then the comment "ends at the end of the last day" would be more correct, because 23:59:59 is the last second of the day, while when setting time to 24:00 it's actually the first second of the next day (you can double check that).

donquixote replied:
A day is an interval. Perhaps we could call it a half-closed interval. So `[start...end[` or `[start...end)` in math notation, or `t >= t_start && t < t_end`. The `00:00:00` moment already belongs to the next day, but every moment nanoseconds before that is from the current day.
So the most semantically correct timestamp we can send is exactly the end of the day. It is not the last moment in the interval, but the upper limit. `sup`, not `max`.
Then it is up to the js what to do with it, to interpret it as an upper limit, not as a max. (but it doesn't make any different in practice I think)

donquixote replied:
> it passes to the next day in the calendar, so if end date is 31st it passes to 1st of next month

this would be ok or not?
the timestamp is only used to calculate the progress bar and status label.

donquixote replied:
btw 23:59:59 is incorrect for summer / winter time switch :)

drishu replied:
not sure about the summer saving but agreed that in this specific use case we can get away with the +1

drishu commented on templates/content/node--oe-project--oe-w-content-banner.html.twig:
```diff
@@ -0,0 +1,17 @@
+  'content': content.oe_teaser,
+  'image': image,
+  'attributes': create_attribute().addClass(['ps-0']),
+  'badges': content.oe_subject ? content.oe_subject|field_value|default([])|map(_item => {
```
next level chaining of filters, but a bit hard 2 read, can we make this more readable for maintainability?

donquixote replied:
would be more readable if we had some custom twig filters.

donquixote replied:
All the parts are still needed to deal with non-array values. And I think this stuff is still better than foreach + merge.
I can change the order, I can insert line breaks.
E.g.
```
  'badges': content.oe_subject ? (content.oe_subject
    |field_value
    |default([])
    |filter(_item => _item|drupal_escape|striptags|trim is not empty)
    |map(_item => {
      'label': _item,
      'rounded_pill': true,
    }))
```

donquixote replied:
or
```
  'badges': content.oe_subject ? (content.oe_subject
    |field_value|default([])
    |filter(_item => _item|drupal_escape|striptags|trim is not empty)
    |map(_item => {
      'label': _item,
      'rounded_pill': true,
    }))
```

donquixote replied:
Actually we can drop the first ternary, because `|field_value` allows null passed to it. The rest will just process an empty array which is ok.

drishu commented on tests/src/Functional/ContentProjectRenderTest.php:
```diff
@@ -4,29 +4,40 @@
 
 namespace Drupal\Tests\oe_whitelabel\Functional;
```
javascript tests belong in a different folder

drishu commented on tests/src/Functional/ContentProjectRenderTest.php:
```diff
@@ -291,4 +320,102 @@ protected function getStorage(string $entity_type_id): EntityStorageInterface {
+   * @param string $end
+   *   End date string.
+   */
+  protected function setProjectDateRange(NodeInterface $node, string $begin, string $end): void {
```
```suggestion
  protected function setProjectDateRangeAndRefresh(NodeInterface $node, string $begin, string $end): void {
```

But honestly a function should do 1 thing and 1 thing only, so the AndSomething should already tell you its nok.
The test is already full of drupalGets, I dont see why it would be an issue to continue, its part of the tests stroy to see you revisit the page.

drishu commented on tests/src/Functional/ContentProjectRenderTest.php:
```diff
@@ -291,4 +320,102 @@ protected function getStorage(string $entity_type_id): EntityStorageInterface {
+  }
+
+  /**
+   * Asserts start and end timestamp in project status section.
```
```suggestion
   * Asserts 'start' and 'end' timestamp in project status section.
```

donquixote replied:
Not sure. The quotes imply that this refers to existing keys or machine names somewhere. Which is not really the case.

drishu replied:
mm, ok

drishu commented on tests/src/Functional/ContentProjectRenderTest.php:
```diff
@@ -291,4 +320,102 @@ protected function getStorage(string $entity_type_id): EntityStorageInterface {
+  /**
+   * Asserts that the d-none class has been removed from project status.
+   */
+  protected function assertProjectStatusRevealed(): void {
```
```suggestion
  protected function assertProjectStatusVisible(): void {
```

drishu commented on tests/src/Functional/ContentProjectRenderTest.php:
```diff
@@ -244,9 +247,35 @@ public function testProjectRendering(): void {
+    ], $description_lists[3]->getHtml());
+
+    // Set a project period that is fully in the past.
+    $this->setProjectDateRange($node, '2019-03-07', '2019-03-21');
```
setProjectDateRange already calls node->save and drupalGet, making below calls a problem. But actually, the fact that setProjectDateRange contains those is wrong, at least the drupalGet, the node save you can keep in there and remove from here.

As you can already see, this is why its so important to have the method names be descriptive to what they do, because then I don't have to check what it does, and you and me both could have spotted this earlier.

drishu reviewed (changes requested):
This is it for now, these are big changes so I'll need to see them implemented before I go further.

donquixote committed:
OEL-1371: Use <article> wrapper, and don't add margin-top.

donquixote committed:
OEL-1371: Don't link the funding programme.

donquixote committed:
OEL-1371: Add function docs in javascript.

donquixote committed:
OEL-1371: Add @todo note because setInterval() only works for up to 10 years duration.

donquixote committed:
OEL-1371: Use const for msBegin / msEnd.

donquixote committed:
OEL-1371: Don't animate the js.

donquixote committed:
OEL-1371: Use formatted dates.

donquixote committed:
OEL-1371: Remove a wrong negation -> hide pie chart if eu budget is NULL.

donquixote committed:
OEL-1371: Simplify project contributions, relying fully on BCL component.

donquixote committed:
OEL-1371: Move the js test into 'FunctionalJavascript' namespace.

donquixote committed:
OEL-1371: Don't save and refresh inside setProjectDateRange().

donquixote committed:
OEL-1371: Rename assertion method -> assertProjectStatusVisible().

donquixote committed:
OEL-1371: Attempt to make chained twig filters more readable.

donquixote committed:
OEL-1371: The ternary expression for badges is not needed.

donquixote committed:
OEL-1371: Add quotes for start/end in comment.

donquixote commented on tests/src/Functional/ContentProjectRenderTest.php:
```diff
@@ -291,4 +320,102 @@ protected function getStorage(string $entity_type_id): EntityStorageInterface {
+   * @param string $end
+   *   Expected end date string as 'Y-m-d H:i:s'.
+   */
+  protected function assertProjectStatusTimestampsAsDateStrings(string $begin, string $end): void {
```
Btw here I use a hard-coded timezone, but in `assertTimestampAsDateString()` I have a timezone param.
I'd say the rationale is that `assertTimestampAsDateString()` is designed as a more generic function, even if used only once.

drishu replied:
I dont mind

donquixote committed:
OEL-1371: Move variable initialization, and merge ifs.

donquixote committed:
OEL-1371: Restore server-side calculation of status and progress.

donquixote committed:
OEL-1371: Use datetime_testing to make tests fail with js disabled.

drishu commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,147 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  $node = $variables['node'];
+  /** @var \Drupal\datetime_range\Plugin\Field\FieldType\DateRangeItem|null $date_range_item */
+  $date_range_item = $node->get('oe_project_dates')->first();
+  if ($date_range_item === NULL || !$date_range_item->value || !$date_range_item->end_value) {
```
I don't mind the thoroughness, but just FYI the field has validation so either both are empty or both have value

drishu replied:
up 2 u

donquixote replied:
The notorious question about defensive code in Drupal.
I don't trust these outside world expectations enough to just ignore them, but I also don't trust them enough to log an error.
So I end up with silent suppression of unexpected cases.

donquixote replied:
Also depends if a website will change the config.
In that case we could log a warning or just ignore and not render the thing.

drishu commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,147 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  $t_end = $get_timestamp($date_range_item->end_value . ' +1 day');
+
+  if ($t_start >= $t_end) {
+    // Cannot show a progress bar if start and end timestamp are the same.
```
comment is inaccurate but not a blocker, we are behind, so..

drishu commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,62 @@
+          $element.addClass(colorClasses[status]);
+          // Disable css transition.
+          // It looks bad for the initial setting, and afterwards it does not help.
+          $element.css('transition', false);
```
I didn't notice any issue, dont think we should fiddle with these, rather report, but not a blocker, seems to work the same with or without. If you found issue create a ticket, and once fixed, remove this "fix"

donquixote replied:
True, no longer necessary.

donquixote replied:
The transition would only go from cached value to current value.
But it does not happen anyway because the element is hidden on page load.
So I remove this code.

drishu commented on tests/src/FunctionalJavascript/ContentProjectRenderTest.php:
```diff
@@ -0,0 +1,441 @@
+
+    $this->assertProjectStatusTimestampsAsDateStrings('2019-03-07 00:00:00', '2019-03-22 00:00:00');
+    $this->assertProjectStatusVisible();
+    $this->assertProjectStatus('bg-dark', 'Closed');
```
the project date range is in 2019
the current date was set to 2017
2017 < 2019 that's planned
so why does it say closed ?

I think the issue is there that the datetime_testing is not used correctly, and the current date is actually always now/today, that would explain the passing tests.
See https://github.com/openeuropa/oe_theme/blob/be9ee44e38b28f13acf138fc16934bf3a7c756df/tests/src/Functional/ContentCallForTendersRenderTest.php for usage example

donquixote replied:
Selenium uses the current date, the server uses the manipulated date.
The tests would fail without javascript (and they do), but they pass with javascript.
That's the point of the ->setTime().
I should add a comment.

I also wanted to add assertions for non-js situation, but did not find how to access the unmodified html.

drishu commented on tests/src/FunctionalJavascript/ContentProjectRenderTest.php:
```diff
@@ -0,0 +1,441 @@
+    $this->assertProjectProgress(100);
+
+    // Set a project period that is ongoing.
+    $this->setProjectDateRange($node, '-5 day', '+15 days');
```
I don't like this relative value, because its was not clear for me from the beginning if this is -5 days from the prev 2019 or -5 days from today 2022. Had to go to the function and check implementation, then scroll back to where I was.

donquixote replied:
True, now it is unclear because we manipulated server time.
We do need dates relative to current time, because this is what Selenium uses.

drishu commented on tests/src/FunctionalJavascript/ContentProjectRenderTest.php:
```diff
@@ -0,0 +1,441 @@
+
+    $this->assertProjectStatusVisible();
+    $this->assertProjectStatus('bg-secondary', 'Planned');
+    $this->assertProjectProgress(0);
```
all this testing is fine, but I think it would be useful, since its a js test to also have:

- a node with a fixed date range
- visit the node assert what needs to be, ex status ongoing
- change the current date to past (but not the nodes dates as now), assert status changed to planning
- change the current date to future, assert status changed to closed

donquixote replied:
not possible because we cannot manipulate Selenium date (or I don't know how).

drishu commented on tests/src/FunctionalJavascript/ContentProjectRenderTest.php:
```diff
@@ -0,0 +1,441 @@
+   *   Overridden time object.
+   */
+  protected function getTestTimeObject(): TestTimeInterface {
+    $time = \Drupal::time();
```
try with 
```
$static_time = new DrupalDateTime('2020-03-02 14:00:00', DateTimeItemInterface::STORAGE_TIMEZONE);
    $this->freezeTime($static_time);
```

donquixote committed:
OEL-1371: Explain and enhance defensive code in preprocess.

donquixote committed:
OEL-1371: Don't remove CSS transition on progress bar.

The transition would be pointless, but it does not happen anyway if the element is hidden on page load.

donquixote committed:
OEL-1371: Drop the datetime_testing in test.

donquixote committed:
OEL-1371: Drop server-side calculation of status and progress. Back to placeholder values.

donquixote committed:
OEL-1371: Change placeholder values to be more neutral.

donquixote committed:
OEL-1371: Use non-empty badge label to make it show up.

brummbar commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,147 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+ */
+function _oe_whitelabel_extra_project_preprocess_contributions(array &$variables): void {
+  $legend_items = [];
+  foreach ([
+    'oe_project_budget' => 'bg-gray-400',
+    'oe_project_budget_eu' => 'bg-primary',
+  ] as $field_name => $bg_class) {
```
I'm really not a fan of this notation, it makes it harder to read. Extract the variable from it.

donquixote replied:
Ok if you say so :)

brummbar replied:
It's literally point 5 of what you linked here https://www.drupal.org/project/coding_standards/issues/1539712 😄

brummbar commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,147 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  $overall_budget = $node->get('oe_project_budget')->value;
+  $eu_budget = $node->get('oe_project_budget_eu')->value;
+
+  if ($overall_budget === NULL || $eu_budget === NULL
+    || $overall_budget <= 0 || $eu_budget < 0
+    || $overall_budget < $eu_budget
```
```suggestion
  if (
    $overall_budget === NULL ||
    $eu_budget === NULL ||
    $overall_budget <= 0 ||
    $overall_budget < $eu_budget
```

donquixote replied:
There is a discussion with a proposed format for this, not yet accepted though.
https://www.drupal.org/project/coding_standards/issues/1539712
The direction is to start each new line with the respective operator.
I support this general direction. I like to see lines starting with the operator.

In this case I cheated a bit to have similar conditions on the same line. But I am ok to split these.

In my own projects I sometimes did this to avoid the first line looking different from the other lines, hoping that the "compiler" will optimize it away. But it is unorthodox and I don't suggest doing it:

```php
if (false
  || $condition1
  || $condition2
) {..}
```

Do we have in-house conventions that the operator should be at the end of a line?

brummbar replied:
Uhm not really, thanks for pointing that issue. I usually use mine above because it aligns all the text, allows to comment out lines without breaking, and it's similar to multi-line function invocations.
Let's keep yours, but split each || in a new line.
Also note that my suggestion drops the check for $eu_budget < 0 as we are checking that for overall budget already, so if overall budget is major than 0 and eu budget is major than overall, it implies that eu is also major than 0. Am I wrong?

donquixote replied:
```php
$eu_budget = -10;
$overall_budget = 50;

if ($overall_budget === NULL  # false
    || $eu_budget === NULL  # false
    || $overall_budget <= 0  # false
    || $eu_budget < 0  # true
    || $overall_budget < $eu_budget  # false
  ) {
    // No pie chart can be drawn with these values.
    return;
  }
```
The condition for `$eu_budget < 0` is the only one that is `true`.
Without the condition, the chart is shown. With the condition, the chart is not shown.

(It is another question whether we should prevent this with validation. But I don't like to rely too much on outside constraints for my code.)

brummbar replied:
Yup you are correct indeed.

brummbar commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,147 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  }
+
+  $ratio_01 = max(0, min(1, $eu_budget / $overall_budget));
+  $percent = (int) round($ratio_01 * 10) * 10;
```
I don't understand the complexity here in the first line. We know that overall and eu budgets are > 0, so we will have always a number between 0 and 1 out of their fraction.

brummbar replied:
We are also missing a comment here to understand that the ratios work only 10 by 10.

brummbar commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,59 @@
+        const msEnd = $element.data('end-timestamp') * 1000;
+        const statusLabels = $element.data('status-labels').split('|');
+        const msNow = Date.now();
+        const status = (msNow >= msBegin) + (msNow > msEnd);
```
Small explanation that 0 is first status, 1 is second and so on.

brummbar commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,59 @@
+          const $element = $(this);
+          $element.removeClass(colorClasses);
+          $element.addClass(colorClasses[status]);
+          if (statusLabels) {
```
How come we wrap this with an if()?

donquixote replied:
Probably leftover. There is no reason now why it would be empty.

brummbar commented on templates/content/node--oe-project--full.html.twig:
```diff
@@ -4,46 +4,50 @@
+  {% if contributions_args is defined %}
+    {% include '@oe-bcl/project-status/project-contributions' with contributions_args only %}
+  {% endif %}
+  {{ content.group_project_details|filter((subgroup, key) => key|first != '#') }}
```
What are we doing here? It looks too much cruft.

donquixote replied:
This is element_children() for poor people.
We want to render each element in the group, but not the group wrapper. We want the wrapper around the entire project details including the status/progress and the contributions.

I would much prefer to have custom twig filters in oe_bootstrap_theme or elsewhere, see https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-1447.
Or a custom field group formatter "bare" that does not render any wrappers.

Alternatively I could render each item in the group separately.

brummbar replied:
Not good to skip wrappers, because cache information can be there in some scenarios. Much better to use a bare template.

donquixote replied:
I think for field groups all the relevant cache tags sit in the fields, so it would be safe like this. But I could be wrong :)

But it makes me wonder if the twig filters `|field_value` and `|field_label` from https://www.drupal.org/project/twig_field_value are safe for caching.
The module page has a cache warning about field_raw and field_target_entity but not about field_value and field_label.

I need to learn a bit more how cache tags are processed during the render process.

For the bare template: I don't like adding more and more specific templates. But something like this could work (this one is for a field, not a group):
```twig
content.body|merge({'#theme': 'field__bare'})
```

donquixote replied:
I think we are safe here, as said. I will leave this one unchanged for now.

brummbar replied:
Can't we just change the template suggestion in the node full preprocess?

donquixote replied:
I did some debugging, and found that field group elements don't contain `'#cache'` entries.

Field group elements are populated in a preprocess on the node/entity itself, which simply moves fields around.

One thing we _could_ do is this:

```twig
{{ content.group_project_details|merge({'#theme_wrappers': []}) }}
```

This way the wrapper div would not be rendered, but the group element would still go through the render pipeline.
Perhaps in addition we also want to remove `'#attached'` data.
Obviously any processing applied by the field group formatter in the preprocess phase already happened before it hits the template.

A cleaner option would be to have a special field group formatter that does not apply any wrappers. See https://www.drupal.org/project/field_group/issues/3279137.
It seems straightforward, we already have one field group formatter in oe_whitelabel.

I am not sure if I want to do this in the scope of this issue.

brummbar commented on templates/content/node--oe-project--oe-w-content-banner.html.twig:
```diff
@@ -0,0 +1,20 @@
+  'content': content.oe_teaser,
+  'image': image,
+  'attributes': create_attribute().addClass(['ps-0']),
+  'badges': content.oe_subject
+    |field_value|default([])
+    |filter(_item => _item|drupal_escape|striptags|trim is not empty)
+    |map(_item => {
+      'label': _item,
+      'rounded_pill': true,
+    }),
```
This also looks incredibly complex, is there a need for this?

donquixote replied:
I don't like it, but I also don't like the for loops with array merge to add new items.

So:
- `|field_value` is like `element_children()`, but unfortunately it returns NULL sometimes. Both `|filter` and `|map` below require an array as parameter, so we need `|default([])`.
- The `|filter()` cleans out field items that are empty when rendered. Perhaps we can skip this step.
- `|map` does the actual array structure conversion.

I will test if empty leftover field items are a thing. If not, I will drop `|filter()`.

brummbar replied:
I think we need to fix in general how we do this, because again cache information attached to the "empty" values can be lost.
Let's keep it for now but revisit later.

donquixote replied:
Actually `|drupal_escape` triggers rendering.
On the other hand, `|field_value` does not.
But ok let's keep. If we have to change this here, there are plenty of other places where we do something that will have the same effect.

brummbar replied:
"Hey this can lead to a bug, but we did it a bit everywhere so let's ignore it". Sounds like a flawed approach to me :)
But anyway I'll make a follow-up to replace field_value with our own filter as we discussed.

donquixote replied:
I did some digging.
Indeed some formatters add cache info directly to the `$elements` array, which is lost if we pick the children.
This will be lost with `|field_value`, but also when using a `for` loop to pick field items.

Let's indeed fix this with a new issue.

brummbar commented on tests/src/FunctionalJavascript/ContentProjectRenderTest.php:
```diff
@@ -0,0 +1,419 @@
+    $this->setUpSparql();
+
+    Role::load(RoleInterface::ANONYMOUS_ID)
+      ->grantPermission('bypass node access')
```
Uhm can't we have a more normal permission? like the others below.

donquixote replied:
Turns out we don't need a perm at all to see the node.

donquixote replied:
Now I notice where this is coming from:
`Drupal\Tests\oe_whitelabel\Kernel\ContentRenderTestBase::setUp()`.
We also don't need it here.

brummbar commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,147 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  }
+
+  // Dates only store the date, not the time.
+  // Use the site-wide configured timezone, not a user-specific timezone.
```
Did we decide on this? I remember something but not precisely. What are the drawbacks of using date_default_timezone_get() and thus a user-specific timezone, if defined?

brummbar replied:
This is actually not aligned with the dates we are rendering below. Drupal doesn't take timezone into consideration when rendering date only fields.
This makes everything harder, since we have the dilemma of "is the date referring the day for the site timezone or my day" or "why this progress is not reflecting the other website dates which are using my user timezone"
or "why the event is over while day X is still in progress".
I'd say we remove this for now completely. Let's keep the utc dates, do something on the lines of
```
  $start = $date_range_item->start_date->setTime(0, 0, 0, 0);
  $end = $date_range_item->end_date->setTime(23, 59, 59, 0);
  $t_start = $start->getTimeStamp();
  $t_end = $end->getTimeStamp();
```
And that's it. We will follow up this to see how to tackle.

brummbar replied:
I am very conflicted about this, also DGCOMM is planning to add timezones to this field so it can be made clear.

donquixote replied:
The scenario is this:
- Site timezone is Brussels,BE.
- A logged-in user has a custom timezone in Sydney,Australia. Depending on site settings, the php timezone will be Sydney, not Brussels, when the request is processed.
- An anonymous user from Sydney has their own local time, but Drupal does not know about it. The php timezone will be Brussels, even though the anon user is in Sydney.

The events only store a date string without time.
We assume that the events happen in Brussels, so they begin and end at 00:00 in Brussels timezone.

With the current calculation, for both of the users from Sydney, the event would show up as "Planned" in the morning, and "Ongoing" in the evening.
If a user from California, one from Brussels, and one from Sydney, logged in or not, have a phone call while visiting the website, they will see the same project status.

If we use user timezone (which will happen if we use php timezone), then a user from California would see a different event status than a user from Sydney, and the anonymous visitor would see a different status than the logged-in user from Sydney.

> This is actually not aligned with the dates we are rendering below.

Indeed, the status would not conform with the date a user would see in their local clock.
However, this is already the case for anon visitors.

> I'd say we remove this for now completely. Let's keep the utc dates, do something on the lines of

I think your code will just use php timezone, which can be the configured timezone or the user timezone.
But let's say we use UTC - so an event starts and ends at 00:00 UTC.
I don't think this is what clients would want. I think most of them will want 00:00 Brussels time. And for those who don't, UTC is going to be just as wrong as Brussels.

I think 00:00 Brussels (or configured site default timezone) is the best we can do with the given information, every alternative is worse.

> I am very conflicted about this, also DGCOMM is planning to add timezones to this field so it can be made clear.

Once this happens, we can review this logic :)

brummbar replied:
I think what you suggested is very much in line with basically saying "The date entered is in the website timezone, so the project is in progress based on that timezone". Which I like, it was one of the "I'm confused" scenarios.
Let's see @drishu opinion, for me is a keep.

drishu replied:
tested again, I agree its good, the main thing that gives me confidence is that the javascript Date.now() is the same no matter the users location or system settings, so it will be the same calculation

brummbar commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,59 @@
+        const msNow = Date.now();
+        const status = (msNow >= msBegin) + (msNow > msEnd);
+        const progress01 = Math.max(0, Math.min(1, (msNow - msBegin) / (msEnd - msBegin)));
+        const percent = Math.round(progress01 * 200) / 2;
```
How come this 2 factor? Can't it be * 100 without the / 2? We need documentation to understand the reasons.

donquixote replied:
Very made-up and artificial, perhaps silly.
Arbitrary long precision looks nasty if the aria attributes are shown somewhere or read in a screen reader.
With 200 steps we are likely to be below pixel width on most displays.
But 100 steps would also work, and raise fewer questions.

brummbar replied:
Uhm I still don't get it much, since we are rounding anyway... Let's remove it indeed.

donquixote replied:
I am now rounding to 1%.
But to explain my idea, my feeling was `37.5%` looks ok but `37.1%` already looks too fine-grained.
Now the new rationale is `37%` looks ok but `37.5%` looks too fine-grained.

These attributes are for accesibility, but I can only speculate how they may be perceived in practice.

brummbar commented on resources/js/oe_whitelabel.project_status.js:
```diff
@@ -0,0 +1,59 @@
+        const statusLabels = $element.data('status-labels').split('|');
+        const msNow = Date.now();
+        const status = (msNow >= msBegin) + (msNow > msEnd);
+        const progress01 = Math.max(0, Math.min(1, (msNow - msBegin) / (msEnd - msBegin)));
```
Drop a line here. UTC timestamps can be negative if before 1970, so we need to make sure that the current progress stays between 0 and 1(00)%.

donquixote replied:
This has nothing to do with 1970.
The raw progress will be < 0 if the event is in the future. It will be > 1 if the event is in the past.
This would be the same even if the current date and the event are before 1970.

The style and attribute values for the progress bar has to in the range of 0..100 so we must clamp it.

brummbar replied:
Yep, 1970 is just one of the scenarios. The important thing is that we need to keep between 0 and 100. Write that in the comment.

brummbar commented on modules/oe_whitelabel_extra_project/oe_whitelabel_extra_project.module:
```diff
@@ -144,5 +160,147 @@ function _oe_whitelabel_extra_project_preprocess_inpage_nav(array &$variables):
+  // Use the site-wide configured timezone, not a user-specific timezone.
+  /* @see \Drupal\system\TimeZoneResolver::getTimeZone() */
+  /** @var string $timezone */
+  $timezone = \Drupal::config('system.date')->get('timezone.default') ?? 'UTC';
```
We forgot to add this config cache information to the render array.

donquixote committed:
OEL-1371: Invalidate cache when site timezone changes.

donquixote committed:
OEL-1371: No multiline array as foreach value.

donquixote committed:
OEL-1371: Each condition part on a separate line.

donquixote committed:
OEL-1371: Drop unnecessary clamping for contribution calculation.

donquixote committed:
OEL-1371: Explain rounding to 10% steps.

donquixote committed:
OEL-1371: Explain status index and progress in js.

donquixote committed:
OEL-1371: Round to 1% instead of 0.5%, and explain why.

donquixote committed:
OEL-1371: No need to have if() for statusLabels.

donquixote committed:
OEL-1371: Remove unnecessary permission in test.

donquixote committed:
OEL-1371: Also remove 'bypass node access' in ContentRenderTestBase.

drishu reviewed (approved):
Changes are ok, approved
